jax/numpy/lax_numpy.py

```diff
@@ -4222,36 +4221,34 @@ def _quantile(a, q, axis, interpolation, keepdims, squash_nans):
 
 
 @partial(jit, static_argnums=2)
-@partial(vectorize, excluded={0, 2})
-def _searchsorted(a, v, side):
-  op = operator.le if side == 'left' else operator.lt
-
-  def cond_fun(state):
-    start, stop = state
-    return stop - start > 1
-
-  def body_fun(state):
-    start, stop = state
-    mid = (start + stop) // 2
-    go_left = op(v, a[mid])
-    return where(go_left, start, mid), where(go_left, mid, stop)
-
-  result = lax.while_loop(cond_fun, body_fun, (0, a.shape[0]))
-  return where(op(v, a[0]), 0, result[1])
-
+def _searchsorted(a, v, side='left'):
+  op = operator.lt if side == "right" else operator.le
+
+  def fscan(carry, x):
+    low, high = carry
+    mid = (low + high) // 2
+    mask = op(v, a[mid])
+    low = where(mask, low, mid)
+    high = where(mask, mid, high)
+    return (low, high), x
+
+  low = zeros(len(v), dtype=dtypes.canonicalize_dtype(int_))
+  high = full(len(v), len(a), dtype=dtypes.canonicalize_dtype(int_))
+  n_levels = int(np.ceil(np.log2(len(a) + 1)))
+  (low, high), _ = lax.scan(fscan, (low, high), None, length=n_levels)
```

Comment:
I _think_ you could write this as a `fori_loop` and get the same performance, or even a `while_loop` _if_ you ensure the `cond_fun` has a special form of `lambda state: state[0] < n_levels` where the first component of `state` is the loop iteration counter. (It's not important that it's the first element, but I think the expression of the `cond_fun` being like `i < N` may be important.) That should trigger the fixed-trip-count XLA optimization that I'm guessing may have been missing from the original implementation.

Using `scan` always triggers this optimization, so it's not a bad idea to use it here, though it is a bit weird not to use the scanned-over input/output.

Reply:
I'm going to give this a try this moring.

Reply:
Done – I re-expressed this in terms of a `fori_loop`. Benchmarks remain largely unchanged.